src/Set.mo

```diff
@@ -758,6 +757,102 @@ module {
+  /// import Iter "mo:base/Iter";
+  /// import Debug "mo:base/Debug";
+  ///
+  /// persistent actor {
+  ///   let set = Set.fromIter(Iter.fromArray([1, 2, 3]), Nat.compare);
+  ///   let iter = Iter.fromArray([3, 4, 5]);
+  ///   Set.addAll(set, Nat.compare, iter);
+  ///   Debug.print(debug_show(Iter.toArray(Set.values(set)))); // => [1, 2, 3, 4, 5]
+  /// }
+  /// ```
+  ///
+  /// Runtime: `O(m * log(n))`.
+  /// Space: `O(1)` retained memory plus garbage, see the note below.
+  /// where `m` and `n` denote the number of elements in `set` and `iter`, respectively,
+  /// and assuming that the `compare` function implements an `O(1)` comparison.
+  public func addAll<T>(set : Set<T>, compare : (T, T) -> Order.Order, iter : Types.Iter<T>) {
+    for (element in iter) {
+      add(set, compare, element)
+    }
+  };
+
+  /// Deletes all values in `iter` from the specified `set`.
+  /// This is different from `Set.diff()` in that the function will trap if
+  /// `iter` contains an element that is not in `set`.
+  ///
+  /// Example:
+  /// ```motoko
+  /// import Set "mo:base/Set";
+  /// import Nat "mo:base/Nat";
+  /// import Iter "mo:base/Iter";
+  /// import Debug "mo:base/Debug";
+  ///
+  /// persistent actor {
+  ///   let set = Set.fromIter(Iter.fromArray([0, 1, 2]), Nat.compare);
+  ///   let iter = Iter.fromArray([0, 2]);
+  ///   Set.deleteAll(set, Nat.compare, iter);
+  ///   Debug.print(debug_show(Iter.toArray(Set.values(set)))); // => [1]
+  /// }
+  /// ```
+  ///
+  /// Runtime: `O(m * log(n))`.
+  /// Space: `O(1)` retained memory plus garbage, see the note below.
+  /// where `m` and `n` denote the number of elements in `set` and `iter`, respectively,
+  /// and assuming that the `compare` function implements an `O(1)` comparison.
+  public func deleteAll<T>(set : Set<T>, compare : (T, T) -> Order.Order, iter : Types.Iter<T>) {
+    for (element in iter) {
+      delete(set, compare, element)
+    }
+  };
+
+  /// Removes all values in `set` aside from those contained in `iter`.
+  /// This is equivalent to `Set.intersect()` but modifies the set in place.
+  ///
+  /// Example:
+  /// ```motoko
+  /// import Set "mo:base/Set";
+  /// import Nat "mo:base/Nat";
+  /// import Iter "mo:base/Iter";
+  /// import Debug "mo:base/Debug";
+  ///
+  /// persistent actor {
+  ///   let set = Set.fromIter(Iter.fromArray([1, 2, 3]), Nat.compare);
+  ///   let iter = Iter.fromArray([3, 4, 5]);
+  ///   Set.retainAll(set, Nat.compare, iter);
+  ///   Debug.print(debug_show(Iter.toArray(Set.values(difference)))); // => [1, 2]
+  /// }
+  /// ```
+  ///
+  /// Runtime: `O(m * log(n))`.
+  /// Space: `O(1)` retained memory plus garbage, see the note below.
+  /// where `m` and `n` denote the number of values in `set` and `iter`, respectively,
+  /// and assuming that the `compare` function implements an `O(1)` comparison.
+  public func retainAll<T>(set : Set<T>, compare : (T, T) -> Order.Order, iter : Types.Iter<T>) {
+    // TODO: optimize?
```

Comment:
Yeah, this looks quadratic to me.

a simple improvement might be to construct a local Set from iter so you can use contains not linear search for testing common membership.
Also  setList could just be created as an ordinary array since we know the size of set up-front. Probably cheaper.

Reply:
How about we just remove `retainAll()` and then include `*InPlace()` functions which take two Set values?

Reply:
I noticed that Rust, Kotlin, and a few other languages use a predicate instead of collection as an argument to this function, which seems a lot better in general. I'll update the PR to go with this, and we can revise from there.

Reply:
Updated, although it's a bit wonky having to pass both the `compare` and `predicate` functions. Perhaps this is another reason for prioritizing dataclasses or another language feature to remove those as soon as possible.